crdt/sequence.py

```diff
@@ -0,0 +1,247 @@
+
+        # Path the sequence using the new operations
+        for op in patch_log:
+            op.do(self.sequence)
+
+        # Merge the two operation logs
+        self.operations = self.operations.merge(other.operations)
+        return self
+
+    def get(self):
+        """
+        Returns the sorted list of payloads in the sequence.
+        """
+        return [obj.operation.payload for obj in self.sequence if not obj.tombstone]
+
+class Object():
+    """
+    An Object represents a single item in a Sequence.
+    """
+    
+    def __init__(self, operation):
+        self.operation = operation
+        self.tombstone = False
+
+    def print(self):
+        """
+        Prints the object to stdout.
+        """
+        print("owner: {}, payload: {}, tombstone: {}".format(self.operation.owner.print(), self.operation.payload, self.tombstone))
+
+class OpId():
+    """
+    An OpId represents a unique identifier for an Operation.
+    """
+
+    def __init__(self, name, id):
+        self.name = name
+        self.id = id
+
+    def compare(self, other):
+        if self.id < other.id:
+            return -1
+        elif self.id > other.id:
+            return 1
+        else:
+            # Conflicts are resolved arbitrarily by comparing the owner IDs
+            # TODO: We could handle "forks" of the operation log and allow for multiple
+            # histories to be valid.
+            if self.name < other.name:
+                return -1
+            elif self.name > other.name:
+                return 1
+            else:
+                return 0
+
+    def is_earlier(self, other):
+        """
+        Returns true if this OpId is earlier than the other OpId.
+        """
+        return self.compare(other) < 0
+
+    def __eq__(self, other):
+        if not isinstance(other, OpId):
+            return False
+        return self.name == other.name and self.id == other.id
+
+    def print(self):
+        """
+        Prints the OpId to stdout.
+        """
+        return "name: {}, id: {}".format(self.name, self.id)
+
+class OperationType(Enum):
+    """
+    Enum for operations.
+    """
+    INSERT = 0
+    REMOVE = 1
+
+class Operation():
+    """
+    An Operation represents a point-in-time change to a Sequence. Every Operation
+    has a globally-unique ID and references a previous Operation, enabling an order of
+    Operations to be imposed.
+    """
+
+    def __init__(self, owner, action, target, payload=None):
+        if action not in OperationType:
+            raise ValueError("Invalid operation type")
+        self.owner = owner
+        self.action = action
+        self.target = target
+        self.payload = payload
+
+    def do(self, objects):
+        """
+        Applies this Operation to an ordered list of objects.
+        """
+        # Iterate through the sequence to find the index of the target object
+        stomp = False
+        if self.target is None:
```

Comment:
Not sure I understand when self.target would be `None` (vs not) as it is defined in the `__init__` discussion above ("the operation that this operation is performed on") -- could you add an inline comment to clarify?

Reply:
When inserting to a sequence with no prior operations applied or at the end there is no target object, because there is nothing to reference against. The current way of handling this is to make `target=None` a switch which means "insert at the end". I've added that info to the docstring.